### PR TITLE
bpo-29183: Fix double exceptions in wsgiref.handlers.BaseHandler

### PR DESCRIPTION
**Pull request by berkerpeksag**
https://bugs.python.org/issue29183

### CODE REVIEW

File: Lib/wsgiref/handlers.py [hunk 1]
```diff
@@ -179,8 +179,14 @@ def finish_response(self):
                 for data in self.result:
                     self.write(data)
                 self.finish_content()
-        finally:
-            self.close()
+        except:
+            if hasattr(self.result, 'close'):
+                self.result.close()
+            raise
+        # We only call close() when no exception is raised, because it
```
Comment (jleclanche) [inline Lib/wsgiref/handlers.py]: Shouldn't this be in an `else` block then?

Comment (berkerpeksag) [reply Lib/wsgiref/handlers.py]: What's the advantage of using an `else` clause in this specific case? Would it make code more readable? Would it make diff shorter? I like using `try...else` clauses, I just want to know why do you prefer the `try...else` way.

**Comment by vadmium** [comment]
Revision da0ca97 looks okay to me. I presume without the fix, the test would fail because the _h.run_ call will raise the double exception.

**Comment by berkerpeksag** [comment]
@vadmium thanks for the review! Yes, the test will fail without the fix:

```
======================================================================
ERROR: testDontResetInternalStateOnException (test.test_wsgiref.HandlerTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/berkerpeksag/projects/cpython/Lib/wsgiref/handlers.py", line 138, in run
    self.finish_response()
  File "/Users/berkerpeksag/projects/cpython/Lib/wsgiref/handlers.py", line 184, in finish_response
    self.write(data)
  File "/Users/berkerpeksag/projects/cpython/Lib/wsgiref/handlers.py", line 289, in write
    self.send_headers()
  File "/Users/berkerpeksag/projects/cpython/Lib/wsgiref/handlers.py", line 347, in send_headers
    self.send_preamble()
  File "/Users/berkerpeksag/projects/cpython/Lib/wsgiref/handlers.py", line 267, in send_preamble
    self._write(('HTTP/%s %s\r\n' % (self.http_version,self.status)).encode('iso-8859-1'))
  File "/Users/berkerpeksag/projects/cpython/Lib/wsgiref/handlers.py", line 468, in _write
    result = self.stdout.write(data)
  File "/Users/berkerpeksag/projects/cpython/Lib/test/test_wsgiref.py", line 819, in write
    raise CustomException
test.test_wsgiref.HandlerTests.testDontResetInternalStateOnException.<locals>.CustomException

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/berkerpeksag/projects/cpython/Lib/test/test_wsgiref.py", line 824, in testDontResetInternalStateOnException
    h.run(hello_app)
  File "/Users/berkerpeksag/projects/cpython/Lib/wsgiref/handlers.py", line 145, in run
    self.handle_error()
  File "/Users/berkerpeksag/projects/cpython/Lib/wsgiref/handlers.py", line 383, in handle_error
    self.finish_response()
  File "/Users/berkerpeksag/projects/cpython/Lib/wsgiref/handlers.py", line 184, in finish_response
    self.write(data)
  File "/Users/berkerpeksag/projects/cpython/Lib/wsgiref/handlers.py", line 289, in write
    self.send_headers()
  File "/Users/berkerpeksag/projects/cpython/Lib/wsgiref/handlers.py", line 346, in send_headers
    if not self.origin_server or self.client_is_modern():
  File "/Users/berkerpeksag/projects/cpython/Lib/wsgiref/handlers.py", line 359, in client_is_modern
    return self.environ['SERVER_PROTOCOL'].upper() != 'HTTP/0.9'
TypeError: 'NoneType' object is not subscriptable

----------------------------------------------------------------------
```

**Pull request merged**